docs/configuration.rst
.. _configuration:

Configuration
=============

A number of configuration options can be set in Django project that affect the
identicon generation. Each configuration option comes with a default value
that's used if it's not specified explicitly in project settings.

The application will verify configuration options, and raise an
``ImproperlyConfigured`` exception in case of a problem.

``PYDENTICON_ROWS``
-------------------

Specifies how many *block* rows a generated identicon should have. The value
should be a positive integer.

**Default value:** ``5``

``PYDENTICON_COLUMNS``
----------------------

Specifies how many *block* columns a generated identicon should have. The value
should be a positive integer.

**Default value:** ``5``

``PYDENTICON_WIDTH``
--------------------

Specifies the width of generated identicon images in pixels (without
padding). The value should be a positive integer.

**Default value:** ``200``

``PYDENTICON_HEIGHT``
---------------------

Specifies the height of generated identicon images in pixels (without
padding). The value should be a positive integer.

**Default value:** ``200``

``PYDENTICON_PADDING``
----------------------

Specifies the padding that will be added to the generated identicon image. The
padding is specified as tuple containing 4 elements, where each element is a
positive integer.

Each element of the tuple is used for padding the identicon image along one of
the edges. The order is: *top*, *bottom*, *left*, *right*.

**Default value:** ``(20, 20, 20, 20)``

``PYDENTICON_FORMAT``
---------------------

Specifies the default format of the generated identicons. The value should be a
string. Supported values are:

* ``"png"`` (for PNG images)
* ``"ascii"`` (for ASCII/textual representation of identicon)

**Default value:** ``"png"``

``PYDENTICON_FOREGROUND``
-------------------------

Specifies a list or tuple of foreground colours that should be used when
generating the identicons. Each element of list/tuple should be a string
conformant to colour specification from the `Pillow
<http://pillow.readthedocs.org/en/latest/reference/ImageColor.html>`_ library.

In addition to regular RGB, you can also use the RGBA (RGB with alpha channel)
scheme to introduce partial or complete transparency when specyfing
colour. Alpha channel value ranges between 0 (full transparency) to 255 (no
transparency). For example ``rgba(224,224,224,0)`` will result in fully
transparent background, while ``rgba(224,224,244,128)`` will result in
approximatelly 50% transparency.

**Default value:** ``("rgb(45,79,255)", "rgb(254,180,44)", "rgb(226,121,234)",
"rgb(30,179,253)", "rgb(232,77,65)", "rgb(49,203,115)", "rgb(141,69,170)")``

``PYDENTICON_BACKGROUND``
-------------------------

Specifies a (single) background colour that should be used when generating the
identicons. This should be a string conformant to colour specification from the
`Pillow <http://pillow.readthedocs.org/en/latest/reference/ImageColor.html>`_
library. The value should be a string.

In addition to regular RGB, you can also use the RGBA (RGB with alpha channel)
scheme to introduce partial or complete transparency when specyfing
colour. Alpha channel value ranges between 0 (full transparency) to 255 (no
transparency). For example ``rgba(224,224,224,0)`` will result in fully
transparent foreground, while ``rgba(224,224,244,128)`` will result in
approximatelly 50% transparency. Different foreground colours can have different
values for alpha channel.

**Default value:** ``"rgb(224,224,224)"``

``PYDENTICON_DIGEST``
---------------------

Specifies digest class that should be used for generating the identicons. Digest
class should support accepting a single constructor argument for passing the
data on which the digest will be run. Instances of the class should also support
a single hexdigest() method that should return a digest of passed data as a hex
string. The value should be a callable.

**Default value:** ``hashlib.md5``

``PYDENTICON_INVERT``
---------------------

Specifies whether the background and foreground colour in generated identicons
should be inverted (swapped) or not. The value should be a boolean (``True`` or
``False``).

**Default value:** ``False``


docs/releasenotes.rst
Release notes
=============

0.2
---

Update introduces support for Django 1.8+ and some documentation improvements.

New features:

* `DJPYD-7: Update application for use in Django 1.8, 1.9, and Django 1.10
  <https://projects.majic.rs/django-pydenticon/issues/DJPYD-7>`_

  Minimum requirement for Django is now 1.8.x. Fixes are compatible with Django
  1.9.x and 1.10.x as well.

Enhancements:

* `DJPYD-8: Update Pydenticon requirement to version 0.3
  <https://projects.majic.rs/django-pydenticon/issues/DJPYD-8>`_

  Introduced explicit dependency on Pydenticon 0.3, which also introduces
  ability to handle transparency in PNGs.

0.1
---

Initial release of Django Pydenticon. Implemented features:

* Serving of identicons through designated URL.
* User data for generating identicons passed via URL.
* Sane configuration defaults for identicon generator for zero-configuration.
* Ability to set parameters of generated identicons.
* Ability to override some of the identicon generation attributes per-request
  via *GET* parameters.
* Full documentation covering installation, usage, privacy. API reference
  included as well.


django_pydenticon/settings.py
# Standard Python library imports.
import collections
import hashlib

# Django imports
from django.conf import settings
from django.core.exceptions import ImproperlyConfigured

# Read project settings for Pydenticon, or fall-back to default values.
PYDENTICON_ROWS = getattr(settings, "PYDENTICON_ROWS", 5)
PYDENTICON_COLUMNS = getattr(settings, "PYDENTICON_COLUMNS", 5)
PYDENTICON_WIDTH = getattr(settings, "PYDENTICON_WIDTH", 200)
PYDENTICON_HEIGHT = getattr(settings, "PYDENTICON_HEIGHT", 200)
PYDENTICON_PADDING = getattr(settings, "PYDENTICON_PADDING", (20, 20, 20, 20))
PYDENTICON_FORMAT = getattr(settings, "PYDENTICON_FORMAT", "png")
PYDENTICON_FOREGROUND = getattr(settings, "PYDENTICON_FOREGROUND", ( "rgb(45,79,255)",
                                                                     "rgb(254,180,44)",
                                                                     "rgb(226,121,234)",
                                                                     "rgb(30,179,253)",
                                                                     "rgb(232,77,65)",
                                                                     "rgb(49,203,115)",
                                                                     "rgb(141,69,170)" ))
PYDENTICON_BACKGROUND = getattr(settings, "PYDENTICON_BACKGROUND", "rgb(224,224,224)")
PYDENTICON_DIGEST = getattr(settings, "PYDENTICON_DIGEST", hashlib.md5)
PYDENTICON_INVERT = getattr(settings, "PYDENTICON_INVERT", False)

# Validate the settings.
if not isinstance(PYDENTICON_ROWS, int) or PYDENTICON_ROWS <= 0:
    raise ImproperlyConfigured("Setting PYDENTICON_ROWS must be a positive integer.")

if not isinstance(PYDENTICON_COLUMNS, int) or PYDENTICON_COLUMNS <= 0:
    raise ImproperlyConfigured("Setting PYDENTICON_COLUMNS must be a positive integer.")

if not isinstance(PYDENTICON_WIDTH, int) or PYDENTICON_WIDTH <= 0:
    raise ImproperlyConfigured("Setting PYDENTICON_WIDTH must be a positive integer.")

if not isinstance(PYDENTICON_HEIGHT, int) or PYDENTICON_HEIGHT <= 0:
    raise ImproperlyConfigured("Setting PYDENTICON_HEIGHT must be a positive integer.")

if not all([isinstance(p, int) and p >= 0 for p in PYDENTICON_PADDING]) or len(PYDENTICON_PADDING) != 4:
    raise ImproperlyConfigured("Setting PYDENTICON_PADDING must be a 4-tuple where each element is a non-negative integer.")

if PYDENTICON_FORMAT not in ("png", "ascii"):
    raise ImproperlyConfigured("Setting PYDENTICON_FORMAT must be set to one of: 'png', 'ascii'.")

if not all([isinstance(f, str) for f in PYDENTICON_FOREGROUND]):
    raise ImproperlyConfigured("Setting PYDENTICON_FOREGROUND must be a tuple where each element is string representation of colour.")

if not isinstance(PYDENTICON_BACKGROUND, str):
    raise ImproperlyConfigured("Setting PYDENTICON_BACKGROUND must be a string representation of colour")

if not isinstance(PYDENTICON_DIGEST, collections.Callable):
    raise ImproperlyConfigured("Setting PYDENTICON_DIGEST must be a callable digest (usually from hashlib module).")

if not isinstance(PYDENTICON_INVERT, bool):
    raise ImproperlyConfigured("Setting PYDENTICON_INVERT must be a boolean.")
